File: backend/app/services/ai_evaluator.py
"""
AI回答評価サービス
独自ルールベース手法のみを使用してtext_input問題の回答を意味的に評価
外部ライブラリ不使用 ── 標準的な文字列操作のみで実装
"""
import re
import unicodedata
from typing import Optional


def normalize_text(text: str) -> str:
    """テキストを正規化（全角半角、大文字小文字、空白など）"""
    text = unicodedata.normalize('NFKC', text)
    text = text.strip()
    text = text.lower()
    text = re.sub(r'\s+', ' ', text)
    return text


def levenshtein_distance(s1: str, s2: str) -> int:
    """レーベンシュタイン距離を計算（DP）"""
    if len(s1) < len(s2):
        return levenshtein_distance(s2, s1)

    if len(s2) == 0:
        return len(s1)

    prev_row = list(range(len(s2) + 1))
    for i, c1 in enumerate(s1):
        curr_row = [i + 1]
        for j, c2 in enumerate(s2):
            cost = 0 if c1 == c2 else 1
            curr_row.append(min(
                curr_row[j] + 1,       # insertion
                prev_row[j + 1] + 1,   # deletion
                prev_row[j] + cost,    # substitution
            ))
        prev_row = curr_row

    return prev_row[-1]


def calculate_similarity(s1: str, s2: str) -> float:
    """2つの文字列のレーベンシュタイン距離ベースの類似度 (0.0–1.0)"""
    longer = max(len(s1), len(s2))
    if longer == 0:
        return 1.0
    dist = levenshtein_distance(s1, s2)
    return (longer - dist) / longer


def remove_punctuation(text: str) -> str:
    """日本語・英語の句読点や記号を削除"""
    return re.sub(
        r'[、。，．,!！?？;；:：\'\"\'\'""（）()\[\]【】『』「」\-\s]',
        '', text,
    )


def extract_numbers(text: str) -> list[str]:
    """テキストから数値（整数・小数）を抽出"""
    return re.findall(r'-?\d+\.?\d*', text)


def numbers_match(correct: str, user: str) -> tuple[bool, float]:
    """数値が含まれる回答の数値一致を確認"""
    cn = extract_numbers(correct)
    un = extract_numbers(user)
    if not cn:
        return False, 0.0
    if cn == un:
        return True, 0.95
    return False, 0.0


_NUMBER_CONVERSIONS: dict[str, list[str]] = {
    '0': ['〇', 'ゼロ', 'ぜろ', '零', 'れい'],
    '1': ['一', 'いち', 'ひとつ', '1つ'],
    '2': ['二', 'に', 'ふたつ', '2つ'],
    '3': ['三', 'さん', 'みっつ', '3つ'],
    '4': ['四', 'よん', 'し', 'よっつ', '4つ'],
    '5': ['五', 'ご', 'いつつ', '5つ'],
    '6': ['六', 'ろく', 'むっつ', '6つ'],
    '7': ['七', 'なな', 'しち', 'ななつ', '7つ'],
    '8': ['八', 'はち', 'やっつ', '8つ'],
    '9': ['九', 'きゅう', 'く', 'ここのつ', '9つ'],
    '10': ['十', 'じゅう', 'とお'],
}


def normalize_numbers(text: str) -> str:
    """数字表現を統一（全角→半角、漢数字→アラビア数字）"""
    normalized = re.sub(
        r'[０-９]',
        lambda m: chr(ord(m.group()) - 0xFEE0),
        text,
    )
    for digit, variants in _NUMBER_CONVERSIONS.items():
        for v in variants:
            normalized = normalized.replace(v, digit)
    return normalized


def katakana_to_hiragana(text: str) -> str:
    """カタカナをひらがなに変換"""
    return re.sub(
        r'[\u30A1-\u30F6]',
        lambda m: chr(ord(m.group()) - 0x60),
        text,
    )


def reading_match(correct: str, user: str) -> tuple[bool, float]:
    """カタカナ⇔ひらがなの読み一致チェック"""
    c_hira = katakana_to_hiragana(normalize_text(correct))
    u_hira = katakana_to_hiragana(normalize_text(user))

    if c_hira == u_hira:
        return True, 0.97

    clean_c = remove_punctuation(c_hira)
    clean_u = remove_punctuation(u_hira)
    if clean_c == clean_u:
        return True, 0.96

    has_kanji = lambda s: bool(re.search(r'[\u4E00-\u9FFF]', s))
    is_only_kana = lambda s: bool(re.fullmatch(r'[\u3040-\u309F\u30A0-\u30FF\s]+', s))

    if has_kanji(correct) and is_only_kana(user):
        kanji_count = len(re.findall(r'[\u4E00-\u9FFF]', correct))
        kana_count = len(clean_u)
        if kanji_count * 1.5 <= kana_count <= kanji_count * 5:
            return True, 0.85

    if is_only_kana(correct) and has_kanji(user):
        kanji_count = len(re.findall(r'[\u4E00-\u9FFF]', user))
        kana_count = len(clean_c)
        if kanji_count * 1.5 <= kana_count <= kanji_count * 5:
            return True, 0.85

    return False, 0.0


def contains_match(correct: str, user: str) -> tuple[bool, float]:
    """一方が他方を含んでいるかチェック"""
    clean_c = remove_punctuation(normalize_text(correct))
    clean_u = remove_punctuation(normalize_text(user))
    if not clean_c or not clean_u:
        return False, 0.0

    if clean_c in clean_u or clean_u in clean_c:
        shorter = min(len(clean_c), len(clean_u))
        longer = max(len(clean_c), len(clean_u))
        if shorter / longer >= 0.5:
            return True, 0.85
    return False, 0.0


def word_overlap_match(correct: str, user: str) -> tuple[bool, float]:
    """単語レベルの一致（英語等スペース区切り向け）"""
    cw = [w for w in normalize_text(correct).split() if w]
    uw = [w for w in normalize_text(user).split() if w]
    if not cw or not uw:
        return False, 0.0

    cs, us = set(cw), set(uw)
    inter = cs & us
    precision = len(inter) / len(us)
    recall = len(inter) / len(cs)
    if precision == 0 or recall == 0:
        return False, 0.0

    f1 = 2 * precision * recall / (precision + recall)
    if f1 >= 0.7:
        return True, min(0.9, f1)
    return False, f1


def evaluate_semantic_similarity(
    correct_answer: str,
    user_answer: str,
) -> tuple[bool, float, str]:
    """独自ルールで意味的類似度を評価"""
    norm_c = normalize_text(correct_answer)
    norm_u = normalize_text(user_answer)

    if norm_c == norm_u:
        return True, 1.0, "完全一致！正解です。"

    clean_c = remove_punctuation(norm_c)
    clean_u = remove_punctuation(norm_u)
    if clean_c == clean_u:
        return True, 0.98, "表現は少し異なりますが、正解です！"

    num_c = normalize_numbers(clean_c)
    num_u = normalize_numbers(clean_u)
    if num_c == num_u:
        return True, 0.95, "数字表現が正解です！"

    ok, conf = reading_match(correct_answer, user_answer)
    if ok:
        return True, conf, "読みが正解です！"

    ok, conf = numbers_match(correct_answer, user_answer)
    if ok:
        return True, conf, "数値が正解です！"

    ok, conf = contains_match(correct_answer, user_answer)
    if ok:
        return True, conf, "正解です！"

    ok, conf = word_overlap_match(correct_answer, user_answer)
    if ok:
        return True, conf, "正解です！表現が少し異なります。"

    similarity = calculate_similarity(clean_c, clean_u)

    if similarity >= 0.85:
        return True, 0.9, "ほぼ正解です！わずかな表現の違いがあります。"
    if similarity >= 0.7:
        return True, 0.8, "正解です！表現が少し異なりますが、意味は合っています。"
    if similarity >= 0.55:
        return True, 0.7, "概ね正解です！細かい表現に違いがあります。"
    if similarity >= 0.4:
        return False, 0.4, "惜しい！部分的に正しいですが、違いがあります。"
    return False, 0.2, "残念ながら不正解です。正解と大きく異なります。"


async def evaluate_text_answer(
    question_text: str,
    correct_answer: str,
    user_answer: str,
    explanation: Optional[str] = None,
) -> dict:
    """
    text_input問題の回答を意味的に評価

    Returns:
        {"is_correct": bool, "confidence": float, "feedback": str, "exact_match": bool}
    """
    first_line = correct_answer.split('\n')[0].strip()
    patterns = [p.strip() for p in first_line.split('/')]

    best: dict = {
        "is_correct": False,
        "confidence": 0.0,
        "feedback": "残念ながら不正解です。正解と大きく異なります。",
        "exact_match": False,
    }

    for pattern in patterns:
        if pattern.strip() == user_answer.strip():
            return {
                "is_correct": True,
                "confidence": 1.0,
                "feedback": "完全一致！正解です。",
                "exact_match": True,
            }

        ok, conf, fb = evaluate_semantic_similarity(pattern, user_answer)
        if conf > best["confidence"]:
            best = {
                "is_correct": ok,
                "confidence": conf,
                "feedback": fb,
                "exact_match": False,
            }
        if ok and conf >= 0.7:
            return best

    return best


async def batch_evaluate_answers(evaluations: list[dict]) -> list[dict]:
    """複数の回答を一括評価"""
    results = []
    for ev in evaluations:
        r = await evaluate_text_answer(
            question_text=ev["question_text"],
            correct_answer=ev["correct_answer"],
            user_answer=ev["user_answer"],
            explanation=ev.get("explanation"),
        )
        results.append(r)
    return results

File: frontend/src/utils/aiEvaluator.ts
/**
 * AI回答評価ユーティリティ
 * 独自ルールベース手法のみでtext_input問題の回答を評価
 */

interface EvaluationResult {
  is_correct: boolean;
  confidence: number;
  feedback: string;
  exact_match: boolean;
}

/**
 * 不可視文字（ゼロ幅スペース、BOM等）を除去
 */
function stripInvisible(text: string): string {
  return text.replace(/[\u200B-\u200D\uFEFF\u00AD\u200E\u200F\u2060\u2061-\u2064\u2066-\u2069]/g, '');
}

/**
 * テキストを正規化（全角半角、大文字小文字、空白など）
 */
function normalizeText(text: string): string {
  let normalized = text.normalize('NFKC');
  normalized = stripInvisible(normalized);
  normalized = normalized.trim();
  normalized = normalized.toLowerCase();
  normalized = normalized.replace(/\s+/g, ' ');
  return normalized;
}

/**
 * 2つの文字列の類似度を計算（レーベンシュタイン距離ベース）
 */
function calculateSimilarity(str1: string, str2: string): number {
  const longer = str1.length > str2.length ? str1 : str2;
  const shorter = str1.length > str2.length ? str2 : str1;

  if (longer.length === 0) {
    return 1.0;
  }

  const editDistance = levenshteinDistance(longer, shorter);
  return (longer.length - editDistance) / longer.length;
}

/**
 * レーベンシュタイン距離を計算
 */
function levenshteinDistance(str1: string, str2: string): number {
  const matrix: number[][] = [];

  for (let i = 0; i <= str2.length; i++) {
    matrix[i] = [i];
  }

  for (let j = 0; j <= str1.length; j++) {
    matrix[0][j] = j;
  }

  for (let i = 1; i <= str2.length; i++) {
    for (let j = 1; j <= str1.length; j++) {
      if (str2.charAt(i - 1) === str1.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1, // substitution
          matrix[i][j - 1] + 1,     // insertion
          matrix[i - 1][j] + 1      // deletion
        );
      }
    }
  }

  return matrix[str2.length][str1.length];
}

/**
 * 句読点や記号を削除
 */
function removePunctuation(text: string): string {
  // 日本語・英語の句読点を削除
  const punctuation = /[、。，．,!！?？;；:：'"''""（）()\[\]【】『』「」\-\s]/g;
  return text.replace(punctuation, '');
}

/**
 * テキストから数値を抽出
 */
function extractNumbers(text: string): string[] {
  // 整数と小数を抽出
  const matches = text.match(/-?\d+\.?\d*/g);
  return matches || [];
}

/**
 * 数値が含まれる回答の場合、数値の一致を確認
 */
function numbersMatch(correctAnswer: string, userAnswer: string): { match: boolean; confidence: number } {
  const correctNums = extractNumbers(correctAnswer);
  const userNums = extractNumbers(userAnswer);

  if (correctNums.length === 0) {
    return { match: false, confidence: 0.0 };
  }

  if (JSON.stringify(correctNums) === JSON.stringify(userNums)) {
    return { match: true, confidence: 0.95 };
  }

  return { match: false, confidence: 0.0 };
}

/**
 * カタカナをひらがなに変換
 */
function katakanaToHiragana(text: string): string {
  return text.replace(/[\u30A1-\u30F6]/g, (match) => {
    const chr = match.charCodeAt(0) - 0x60;
    return String.fromCharCode(chr);
  });
}

/**
 * 数字と漢数字・かなの対応表
 */
const numberConversions: { [key: string]: string[] } = {
  '0': ['〇', 'ゼロ', 'ぜろ', '零', 'れい'],
  '1': ['一', 'いち', '１', 'ひとつ', '1つ'],
  '2': ['二', 'に', '２', 'ふたつ', '2つ'],
  '3': ['三', 'さん', '３', 'みっつ', '3つ'],
  '4': ['四', 'よん', 'し', '４', 'よっつ', '4つ'],
  '5': ['五', 'ご', '５', 'いつつ', '5つ'],
  '6': ['六', 'ろく', '６', 'むっつ', '6つ'],
  '7': ['七', 'なな', 'しち', '７', 'ななつ', '7つ'],
  '8': ['八', 'はち', '８', 'やっつ', '8つ'],
  '9': ['九', 'きゅう', 'く', '９', 'ここのつ', '9つ'],
  '10': ['十', 'じゅう', 'とお', '１０'],
};

/**
 * 数字表現の正規化（アラビア数字、漢数字、かな表記の統一）
 */
function normalizeNumbers(text: string): string {
  let normalized = text;

  // 全角数字を半角に変換（NFKC正規化で対応済みだが明示的に）
  normalized = normalized.replace(/[０-９]/g, (s) => String.fromCharCode(s.charCodeAt(0) - 0xFEE0));

  // 漢数字・かな表記をアラビア数字に変換
  for (const [digit, variants] of Object.entries(numberConversions)) {
    for (const variant of variants) {
      // 完全一致または単語境界でのみ置換
      const regex = new RegExp(`\\b${variant}\\b|${variant}(?=[のつ])`, 'g');
      normalized = normalized.replace(regex, digit);
    }
  }

  return normalized;
}

/**
 * 漢字の読み（ひらがな・カタカナ）が一致するかチェック
 */
function readingMatch(correctAnswer: string, userAnswer: string): { match: boolean; confidence: number } {
  const hasKanji = (str: string) => /[\u4E00-\u9FFF]/.test(str);
  const hasKana = (str: string) => /[\u3040-\u309F\u30A0-\u30FF]/.test(str);
  const isOnlyKana = (str: string) => /^[\u3040-\u309F\u30A0-\u30FF\s]+$/.test(str);

  // 両方をひらがなに統一（カタカナのみ変換）
  const correctHiragana = katakanaToHiragana(normalizeText(correctAnswer));
  const userHiragana = katakanaToHiragana(normalizeText(userAnswer));

  // 完全一致（カタカナ⇔ひらがな）
  if (correctHiragana === userHiragana) {
    return { match: true, confidence: 0.97 };
  }

  // 句読点を除いて一致
  const cleanCorrect = removePunctuation(correctHiragana);
  const cleanUser = removePunctuation(userHiragana);

  if (cleanCorrect === cleanUser) {
    return { match: true, confidence: 0.96 };
  }

  // 一方が漢字・もう一方がかなの場合（簡易的な読みチェック）
  // 例: 「東京」vs「とうきょう」, 「東京」vs「トウキョウ」
  if (hasKanji(correctAnswer) && isOnlyKana(userAnswer)) {
    // 正解が漢字含む、回答がかなのみ → 読みの可能性が高い
    // 漢字1文字は通常1.5-4文字のかなになる(例: 「日」→「に」1文字、「本」→「ほん」2文字)
    const kanjiCount = (correctAnswer.match(/[\u4E00-\u9FFF]/g) || []).length;
    const kanaCount = cleanUser.length;

    // 条件: 漢字数 * 1.5 以上、かつ漢字数 * 5 以下
    // 例: 漢字2文字 → 3-10文字のかなが妥当
    if (kanaCount >= Math.floor(kanjiCount * 1.5) && kanaCount <= kanjiCount * 5) {
      return { match: true, confidence: 0.85 };
    }
  }

  if (isOnlyKana(correctAnswer) && hasKanji(userAnswer)) {
    // 正解がかな、回答が漢字 → 逆パターン（例: りんご vs 林檎）
    const kanjiCount = (userAnswer.match(/[\u4E00-\u9FFF]/g) || []).length;
    const kanaCount = cleanCorrect.length;

    // かな数が漢字数 * 1.5 以上、かつ漢字数 * 5 以下なら可能性あり
    if (kanaCount >= Math.floor(kanjiCount * 1.5) && kanaCount <= kanjiCount * 5) {
      return { match: true, confidence: 0.85 };
    }
  }

  return { match: false, confidence: 0.0 };
}

/**
 * 選択肢記号の先頭プレフィックスを除去（例: "A. 東京" → "東京", "A) answer" → "answer"）
 */
function stripOptionPrefix(text: string): string {
  return text.replace(/^[A-Za-z\uFF21-\uFF3A\uFF41-\uFF5A\u30A2\u30A4\u30A6\u30A8\u30AA]\s*[.):\s、．）]\s*/, '').trim();
}

/**
 * 一方が他方を含んでいるかチェック
 */
function containsMatch(correctAnswer: string, userAnswer: string): { match: boolean; confidence: number } {
  const cleanCorrect = removePunctuation(normalizeText(correctAnswer));
  const cleanUser = removePunctuation(normalizeText(userAnswer));

  if (!cleanCorrect || !cleanUser) return { match: false, confidence: 0 };

  if (cleanCorrect.includes(cleanUser) || cleanUser.includes(cleanCorrect)) {
    const shorter = Math.min(cleanCorrect.length, cleanUser.length);
    const longer = Math.max(cleanCorrect.length, cleanUser.length);
    const ratio = shorter / longer;
    if (ratio >= 0.5) {
      return { match: true, confidence: 0.85 };
    }
  }

  return { match: false, confidence: 0 };
}

/**
 * 選択肢記号としてのマッチ（"A" → "A. 東京" の先頭記号と一致）
 */
function optionLabelMatch(correctAnswer: string, userAnswer: string): { match: boolean; confidence: number } {
  const normUser = normalizeText(userAnswer);
  if (normUser.length > 3) return { match: false, confidence: 0 };

  const normCorrect = normalizeText(correctAnswer);
  const optionPattern = /^([a-z\uFF41-\uFF5A\u30A2\u30A4\u30A6\u30A8\u30AA])\s*[.):\s、．）]/;
  const m = normCorrect.match(optionPattern);
  if (m && m[1] === normUser) {
    return { match: true, confidence: 0.92 };
  }
  return { match: false, confidence: 0 };
}

/**
 * 単語レベルの一致をチェック（英語向け）
 */
function wordOverlapMatch(correctAnswer: string, userAnswer: string): { match: boolean; confidence: number } {
  const correctWords = normalizeText(correctAnswer).split(/\s+/).filter(w => w.length > 0);
  const userWords = normalizeText(userAnswer).split(/\s+/).filter(w => w.length > 0);

  if (correctWords.length === 0 || userWords.length === 0) return { match: false, confidence: 0 };

  const correctSet = new Set(correctWords);
  const userSet = new Set(userWords);
  const intersection = correctWords.filter((w, i, arr) => arr.indexOf(w) === i && userSet.has(w));

  const precision = intersection.length / userSet.size;
  const recall = intersection.length / correctSet.size;

  if (precision === 0 || recall === 0) return { match: false, confidence: 0 };
  const f1 = 2 * (precision * recall) / (precision + recall);

  if (f1 >= 0.7) {
    return { match: true, confidence: Math.min(0.9, f1) };
  }

  return { match: false, confidence: f1 };
}

/**
 * 意味的類似度を評価
 */
function evaluateSemanticSimilarity(correctAnswer: string, userAnswer: string, lang: 'en' | 'ja' = 'en'): {
  isCorrect: boolean;
  confidence: number;
  feedback: string;
} {
  const normCorrect = normalizeText(correctAnswer);
  const normUser = normalizeText(userAnswer);

  if (normCorrect === normUser) {
    return {
      isCorrect: true,
      confidence: 1.0,
      feedback: lang === 'ja' ? '完全一致！正解です。' : 'Perfect match!',
    };
  }

  const cleanCorrect = removePunctuation(normCorrect);
  const cleanUser = removePunctuation(normUser);

  if (cleanCorrect === cleanUser) {
    return {
      isCorrect: true,
      confidence: 0.98,
      feedback: lang === 'ja' ? '表現は少し異なりますが、正解です！' : 'Correct! Slightly different expression.',
    };
  }

  const numNormCorrect = normalizeNumbers(cleanCorrect);
  const numNormUser = normalizeNumbers(cleanUser);

  if (numNormCorrect === numNormUser) {
    return {
      isCorrect: true,
      confidence: 0.95,
      feedback: lang === 'ja' ? '数字表現が正解です！' : 'Correct number!',
    };
  }

  const { match: isReadingMatch, confidence: readingConfidence } = readingMatch(correctAnswer, userAnswer);
  if (isReadingMatch) {
    return {
      isCorrect: true,
      confidence: readingConfidence,
      feedback: lang === 'ja' ? '読みが正解です！' : 'Correct reading!',
    };
  }

  const { match: isNumberMatch, confidence: numConfidence } = numbersMatch(correctAnswer, userAnswer);
  if (isNumberMatch) {
    return {
      isCorrect: true,
      confidence: numConfidence,
      feedback: lang === 'ja' ? '数値が正解です！' : 'Correct number!',
    };
  }

  // 選択肢記号マッチ（"A" → "A. 東京" の先頭ラベル一致）
  const { match: isOptionLabel, confidence: optionConf } = optionLabelMatch(correctAnswer, userAnswer);
  if (isOptionLabel) {
    return {
      isCorrect: true,
      confidence: optionConf,
      feedback: lang === 'ja' ? '正解です！' : 'Correct!',
    };
  }

  // プレフィックスを除去して再比較（"A. 東京" vs "東京" → 一致）
  const strippedCorrect = stripOptionPrefix(normalizeText(correctAnswer));
  const strippedUser = stripOptionPrefix(normalizeText(userAnswer));
  if (strippedCorrect && strippedUser && strippedCorrect === strippedUser) {
    return {
      isCorrect: true,
      confidence: 0.93,
      feedback: lang === 'ja' ? '正解です！' : 'Correct!',
    };
  }

  // 包含チェック（一方がもう一方を含む）
  const { match: isContains, confidence: containsConf } = containsMatch(correctAnswer, userAnswer);
  if (isContains) {
    return {
      isCorrect: true,
      confidence: containsConf,
      feedback: lang === 'ja' ? '正解です！' : 'Correct!',
    };
  }

  // 単語レベル一致（英語等スペース区切り言語向け）
  const { match: isWordMatch, confidence: wordConf } = wordOverlapMatch(correctAnswer, userAnswer);
  if (isWordMatch) {
    return {
      isCorrect: true,
      confidence: wordConf,
      feedback: lang === 'ja' ? '正解です！表現が少し異なります。' : 'Correct! Slightly different wording.',
    };
  }

  const similarity = calculateSimilarity(cleanCorrect, cleanUser);

  if (similarity >= 0.85) {
    return {
      isCorrect: true,
      confidence: 0.9,
      feedback: lang === 'ja' ? 'ほぼ正解です！わずかな表現の違いがあります。' : 'Almost perfect! Minor differences.',
    };
  } else if (similarity >= 0.7) {
    return {
      isCorrect: true,
      confidence: 0.8,
      feedback: lang === 'ja' ? '正解です！表現が少し異なりますが、意味は合っています。' : 'Correct! Slightly different but same meaning.',
    };
  } else if (similarity >= 0.55) {
    return {
      isCorrect: true,
      confidence: 0.7,
      feedback: lang === 'ja' ? '概ね正解です！細かい表現に違いがあります。' : 'Mostly correct! Some differences but right meaning.',
    };
  } else if (similarity >= 0.4) {
    return {
      isCorrect: false,
      confidence: 0.4,
      feedback: lang === 'ja' ? '惜しい！部分的に正しいですが、違いがあります。' : 'Close! Partially correct but some differences.',
    };
  } else {
    return {
      isCorrect: false,
      confidence: 0.2,
      feedback: lang === 'ja' ? '残念ながら不正解です。正解と大きく異なります。' : 'Incorrect. Very different from the answer.',
    };
  }
}

/**
 * text_input問題の回答を意味的に評価（クライアントサイド版）
 */
export function evaluateTextAnswer(correctAnswer: string, userAnswer: string, lang: 'en' | 'ja' = 'en'): EvaluationResult {
  console.log('[evaluateTextAnswer] Called with:', {
    correctAnswer,
    userAnswer,
  });

  // 改行で分割し、最初の行を正解として使用（例文などを除外）
  const firstLine = correctAnswer.split('\n')[0].trim();

  // 「/」で区切られた複数の正解パターンに対応
  const basePatterns = firstLine.split('/').map(s => s.trim());

  // プレフィックス除去版も候補に追加（重複排除）
  const correctAnswerPatterns = basePatterns.slice();
  for (const p of basePatterns) {
    const stripped = stripOptionPrefix(p);
    if (stripped && stripped !== p.toLowerCase() && correctAnswerPatterns.indexOf(stripped) === -1) {
      correctAnswerPatterns.push(stripped);
    }
  }
  console.log('[evaluateTextAnswer] Patterns:', correctAnswerPatterns);

  let bestResult: EvaluationResult = {
    is_correct: false,
    confidence: 0,
    feedback: lang === 'ja' ? '残念ながら不正解です。正解と大きく異なります。' : 'Incorrect. Very different from the answer.',
    exact_match: false,
  };

  // すべての正解パターンを試して、最も高い信頼度のものを採用
  for (const pattern of correctAnswerPatterns) {
    console.log('[evaluateTextAnswer] Evaluating pattern:', pattern);

    // 完全一致チェック（正規化前）
    const exactMatch = pattern.trim() === userAnswer.trim();
    console.log('[evaluateTextAnswer] Exact match check:', exactMatch);

    if (exactMatch) {
      const result = {
        is_correct: true,
        confidence: 1.0,
        feedback: lang === 'ja' ? '完全一致！正解です。' : 'Perfect match!',
        exact_match: true,
      };
      console.log('[evaluateTextAnswer] Result (exact match):', result);
      return result;
    }

    // 意味的類似度評価
    const { isCorrect, confidence, feedback } = evaluateSemanticSimilarity(pattern, userAnswer, lang);
    console.log('[evaluateTextAnswer] Semantic evaluation result:', { isCorrect, confidence, feedback });

    // より高い信頼度の結果を保持
    if (confidence > bestResult.confidence) {
      bestResult = {
        is_correct: isCorrect,
        confidence,
        feedback,
        exact_match: false,
      };
      console.log('[evaluateTextAnswer] Updated best result:', bestResult);
    }

    // 正解と判定されたら即座に返す（最適化）
    if (isCorrect && confidence >= 0.7) {
      console.log('[evaluateTextAnswer] Early return with result:', bestResult);
      return bestResult;
    }
  }

  console.log('[evaluateTextAnswer] Final result:', bestResult);
  return bestResult;
}
